File: frontend/src/api/m1.ts
import { apiClient } from "./client";

export type Role = "admin" | "editor" | "viewer";
export type KnowledgeBaseType = "document" | "wiki";

export type User = {
  id: string;
  username: string;
  created_at: string;
};

export type Workspace = {
  id: string;
  name: string;
  created_by: string;
  created_at: string;
};

export type WorkspaceMember = {
  id: string;
  workspace_id: string;
  user: User;
  role: Role;
  created_at: string;
};

export type TokenPair = {
  access_token: string;
  refresh_token: string;
  token_type: "bearer";
  expires_in: number;
};

export type AuthResponse = {
  user: User;
  workspace?: Workspace | null;
  membership?: WorkspaceMember | null;
  tokens: TokenPair;
};

export type OllamaModel = {
  tag: string;
  digest: string;
  capabilities: string[];
  embedding_dim: number | null;
  usable_for_v1: boolean;
  unusable_reason:
    | "network_error"
    | "model_not_found"
    | "not_embedding_model"
    | "dimension_incompatible"
    | "probe_failed"
    | null;
};

export type ChunkingConfig = {
  chunk_size: number;
  chunk_overlap: number;
  strategy: "header_aware";
};

export type WikiConfig = {
  auto_ingest: boolean;
  llm_timeout_seconds?: number;
  llm_max_retries?: number;
  temperature?: number;
};

export type KnowledgeBase = {
  id: string;
  workspace_id: string;
  name: string;
  description: string;
  type: KnowledgeBaseType;
  status: "active" | "building" | "disabled" | "embedding_incompatible";
  embedding_provider: "ollama";
  embedding_model_tag: string;
  embedding_model_digest: string;
  embedding_dim: 1024;
  chunking_config?: ChunkingConfig | null;
  wiki_config?: WikiConfig | null;
  document_count?: number;
  page_count?: number;
  bound_source_kbs?: Array<Pick<KnowledgeBase, "id" | "name" | "type" | "status">>;
  created_at: string;
  updated_at: string;
};

export type LlmConfig = {
  provider: "openai" | "deepseek";
  model: string;
  base_url: string;
  api_key_configured: boolean;
  api_key_masked: string;
  temperature: number;
  max_tokens: number;
  timeout_seconds: number;
  updated_at: string;
};

export type OllamaConfig = {
  base_url: string;
  updated_at: string;
};

export type ModelTestResult = {
  ok: boolean;
  code: string;
  message: string;
  latency_ms: number;
};

export async function register(payload: {
  username: string;
  password: string;
}): Promise<AuthResponse> {
  const response = await apiClient.post<AuthResponse>("/auth/register", payload);
  return response.data;
}

export async function login(payload: {
  username: string;
  password: string;
}): Promise<AuthResponse> {
  const response = await apiClient.post<AuthResponse>("/auth/login", payload);
  return response.data;
}

export async function getCurrentWorkspace(): Promise<Workspace> {
  const response = await apiClient.get<Workspace>("/workspaces/current");
  return response.data;
}

export async function updateCurrentWorkspace(payload: { name: string }): Promise<Workspace> {
  const response = await apiClient.patch<Workspace>("/workspaces/current", payload);
  return response.data;
}

export async function listMembers(): Promise<WorkspaceMember[]> {
  const response = await apiClient.get<{ items: WorkspaceMember[] }>("/workspaces/current/members");
  return response.data.items;
}

export async function addMember(payload: {
  username: string;
  role: Role;
}): Promise<WorkspaceMember> {
  const response = await apiClient.post<WorkspaceMember>("/workspaces/current/members", payload);
  return response.data;
}

export async function updateMemberRole(userId: string, role: Role): Promise<WorkspaceMember> {
  const response = await apiClient.patch<WorkspaceMember>(`/workspaces/current/members/${userId}`, {
    role,
  });
  return response.data;
}

export async function removeMember(userId: string): Promise<void> {
  await apiClient.delete(`/workspaces/current/members/${userId}`);
}

export async function listKnowledgeBases(params: {
  type?: KnowledgeBaseType;
  q?: string;
} = {}): Promise<KnowledgeBase[]> {
  const response = await apiClient.get<{ items: KnowledgeBase[] }>("/kbs", { params });
  return response.data.items;
}

export async function createKnowledgeBase(payload:
  | {
      type: "document";
      name: string;
      description: string;
      embedding_model_tag: string;
      chunking_config?: ChunkingConfig;
    }
  | {
      type: "wiki";
      name: string;
      description: string;
      embedding_model_tag: string;
      source_knowledge_base_ids: string[];
      wiki_config?: WikiConfig;
    }): Promise<KnowledgeBase> {
  const response = await apiClient.post<KnowledgeBase>("/kbs", payload);
  return response.data;
}

export async function updateKnowledgeBase(
  id: string,
  payload: Partial<Pick<KnowledgeBase, "name" | "description" | "status" | "chunking_config" | "wiki_config">>,
): Promise<KnowledgeBase> {
  const response = await apiClient.patch<KnowledgeBase>(`/kbs/${id}`, payload);
  return response.data;
}

export async function getKnowledgeBase(id: string): Promise<KnowledgeBase> {
  const response = await apiClient.get<KnowledgeBase>(`/kbs/${id}`);
  return response.data;
}

export async function deleteKnowledgeBase(id: string): Promise<void> {
  await apiClient.delete(`/kbs/${id}`);
}

export async function bindSourceKnowledgeBase(kbId: string, sourceKbId: string): Promise<void> {
  await apiClient.post(`/kbs/${kbId}/bindings`, { source_kb_id: sourceKbId });
}

export async function unbindSourceKnowledgeBase(kbId: string, sourceKbId: string): Promise<void> {
  await apiClient.delete(`/kbs/${kbId}/bindings/${sourceKbId}`);
}

export async function getLlmConfig(): Promise<LlmConfig> {
  const response = await apiClient.get<LlmConfig>("/admin/llm-config");
  return response.data;
}

export async function updateLlmConfig(
  payload: Omit<LlmConfig, "api_key_configured" | "api_key_masked" | "updated_at"> & {
    api_key?: string;
  },
): Promise<LlmConfig> {
  const response = await apiClient.put<LlmConfig>("/admin/llm-config", payload);
  return response.data;
}

export async function testLlmConfig(): Promise<ModelTestResult> {
  const response = await apiClient.post<ModelTestResult>("/admin/llm-config/test");
  return response.data;
}

export async function getOllamaConfig(): Promise<OllamaConfig> {
  const response = await apiClient.get<OllamaConfig>("/admin/ollama-config");
  return response.data;
}

export async function updateOllamaConfig(base_url: string): Promise<OllamaConfig> {
  const response = await apiClient.put<OllamaConfig>("/admin/ollama-config", { base_url });
  return response.data;
}

export async function listOllamaModels(): Promise<OllamaModel[]> {
  const response = await apiClient.get<{ items: OllamaModel[] }>("/admin/ollama/models");
  return response.data.items;
}

export async function probeOllamaModel(tag: string): Promise<OllamaModel> {
  const response = await apiClient.post<OllamaModel>("/admin/ollama/models/probe", { tag });
  return response.data;
}

File: backend/app/schemas/m1.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

Role = Literal["admin", "editor", "viewer"]
KbType = Literal["document", "wiki"]
KbStatus = Literal["active", "building", "disabled", "embedding_incompatible"]
LlmProvider = Literal["openai", "deepseek"]


class UserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    username: str
    created_at: datetime


class RegisterRequest(BaseModel):
    username: str = Field(min_length=3, max_length=64)
    password: str = Field(min_length=8, max_length=128)


class LoginRequest(BaseModel):
    username: str
    password: str


class RefreshTokenRequest(BaseModel):
    refresh_token: str


class TokenPair(BaseModel):
    access_token: str
    refresh_token: str
    token_type: Literal["bearer"] = "bearer"
    expires_in: int


class WorkspaceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    created_by: str
    created_at: datetime


class WorkspaceUpdateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128)


class WorkspaceMemberOut(BaseModel):
    id: str
    workspace_id: str
    user: UserOut
    role: Role
    created_at: datetime


class AuthResponse(BaseModel):
    user: UserOut
    workspace: WorkspaceOut | None = None
    membership: WorkspaceMemberOut | None = None
    tokens: TokenPair


class MemberCreateRequest(BaseModel):
    username: str = Field(min_length=3, max_length=64)
    role: Role


class MemberRoleUpdateRequest(BaseModel):
    role: Role


class MemberListResponse(BaseModel):
    items: list[WorkspaceMemberOut]


class LlmConfigResponse(BaseModel):
    provider: LlmProvider
    model: str
    base_url: str
    api_key_configured: bool
    api_key_masked: str
    temperature: float
    max_tokens: int
    timeout_seconds: int
    updated_at: datetime


class LlmConfigUpdateRequest(BaseModel):
    provider: LlmProvider
    model: str = Field(min_length=1, max_length=128)
    base_url: HttpUrl
    api_key: str | None = Field(default=None, min_length=1, max_length=4096)
    temperature: float = Field(default=0.7, ge=0, le=2)
    max_tokens: int = Field(default=4096, ge=1)
    timeout_seconds: int = Field(default=60, ge=1)


class ModelTestResult(BaseModel):
    ok: bool
    code: Literal[
        "ok",
        "network_error",
        "auth_error",
        "model_not_found",
        "timeout",
        "invalid_response",
    ]
    message: str
    latency_ms: int


class OllamaConfig(BaseModel):
    base_url: str
    updated_at: datetime


class OllamaConfigUpdateRequest(BaseModel):
    base_url: HttpUrl


class OllamaModelProbeRequest(BaseModel):
    tag: str = Field(min_length=1, max_length=128)


class OllamaModelProbeResult(BaseModel):
    tag: str
    digest: str
    capabilities: list[str]
    embedding_dim: int | None
    usable_for_v1: bool
    unusable_reason: Literal[
        "network_error",
        "model_not_found",
        "not_embedding_model",
        "dimension_incompatible",
        "probe_failed",
    ] | None


class OllamaModelListResponse(BaseModel):
    items: list[OllamaModelProbeResult]


class ChunkingConfig(BaseModel):
    chunk_size: int = Field(default=512, ge=128, le=4096)
    chunk_overlap: int = Field(default=80, ge=0, le=1024)
    strategy: Literal["header_aware"] = "header_aware"


class WikiConfig(BaseModel):
    auto_ingest: bool = False
    llm_timeout_seconds: int = Field(default=60, ge=1)
    llm_max_retries: int = Field(default=3, ge=0, le=5)
    temperature: float = Field(default=0.7, ge=0, le=2)


class SourceKnowledgeBaseCreateRequest(BaseModel):
    type: Literal["document"]
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default="", max_length=2000)
    embedding_model_tag: str
    chunking_config: ChunkingConfig | None = None


class WikiKnowledgeBaseCreateRequest(BaseModel):
    type: Literal["wiki"]
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default="", max_length=2000)
    embedding_model_tag: str
    source_knowledge_base_ids: list[str] = Field(min_length=1)
    wiki_config: WikiConfig | None = None


class KnowledgeBaseUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    status: Literal["active", "disabled"] | None = None
    chunking_config: ChunkingConfig | None = None
    wiki_config: WikiConfig | None = None


class KnowledgeBaseSummary(BaseModel):
    id: str
    name: str
    type: KbType
    status: KbStatus


class KnowledgeBaseOut(BaseModel):
    id: str
    workspace_id: str
    name: str
    description: str
    type: KbType
    status: KbStatus
    embedding_provider: Literal["ollama"]
    embedding_model_tag: str
    embedding_model_digest: str
    embedding_dim: Literal[1024]
    chunking_config: ChunkingConfig | None = None
    wiki_config: WikiConfig | None = None
    document_count: int = 0
    page_count: int = 0
    bound_source_kbs: list[KnowledgeBaseSummary] = []
    created_at: datetime
    updated_at: datetime


class KnowledgeBasePage(BaseModel):
    items: list[KnowledgeBaseOut]
    total: int
    page: int
    page_size: int


class WikiSourceBindingRequest(BaseModel):
    source_kb_id: str


class WikiSourceBindingOut(BaseModel):
    id: str
    wiki_kb_id: str
    source_kb_id: str
    created_at: datetime


class AuditLogOut(BaseModel):
    id: str
    workspace_id: str
    user_id: str | None = None
    action: str
    resource_type: str
    resource_id: str | None = None
    details: dict[str, Any]
    created_at: datetime


class AuditLogPage(BaseModel):
    items: list[AuditLogOut]
    total: int
    page: int
    page_size: int
